Enabeling opty to estimate input time shifts.

chris-konrad opened the pull request:
This is work in progress addressing https://github.com/csu-hmc/opty/issues/224. The code currently does not run without exceptions and added functions are untested. 

See https://github.com/chris-konrad/opty/issues/1 for status and plans.

moorepants commented:
> See https://github.com/chris-konrad/opty/issues/1 for status and plans.

This would be nice if were an issue here, because forks are often temporary or just may disappear at some point.

chris-konrad commented:
> This would be nice if were an issue here, because forks are often temporary or just may disappear at some point.

I added it to the first message in this PR

chris-konrad commented:
I added the new indices of the new timeshift contraint partials to the jacobian indices. This assumes the following layout of the Jacobian. Can you confirm that this is correct, @moorepants? Green is the existing Jacobian, yellow is the extension with timeshift constraints. 

The constraint definition is as above: 
$G_i = [\frac{\partial c_i}{\partial u[i]}, \frac{\partial c_i}{\partial \tau}]$

Creating a new problem with `Problem()` now runs through without exceptions. I did not test it to any extend so far though.

Peter230655 commented:
@chris-konrad : Curiosity: Once this works will opty be able to handle time delayed eoms, like $\dfrac{sx}{dt = f(x(t - \tau), u(x - \tau}$

chris-konrad commented:
@Peter230655: No, unfortunately this will not directly be possible. The implemented approach requires that the timeshifted trajectory is known trajectory.

Peter230655 commented:
> @Peter230655: No, unfortunately this will not directly be possible. The implemented approach requires that the timeshifted trajectory is a known trajectory.

Thanks!
Still good for me: I am mostly a user of sympy mechanics and of opty, playing around with them to pass time in retirement.
This, when implemented will enlarge the playing field for me. :-)

chris-konrad commented:
All previously existing unittests run fine but it seems to break some examples with variable times. This is little surprising as I did not pay any attention to that during implementation. Before addressing these issues, I will focus on testing if the new functions work at all.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1566,3 +1566,51 @@ def test_for_algebraic_eoms():
+        self.eom = sym.Matrix(((x_1(t).diff() - x_2(t),),
+                          (x_2(t).diff() + a_0 * T_0 * x_1(t) 
+                                              + a_1 * T_0 * x_2(t) 
+                                              - T_0 * u(t-t_d),)))
```
Why would u_shift not be added here instead of u(t-t_d)? Shouldn't the eom's not have t_d present?

chris-konrad replied:
This is the eom that I pass to the constructor of `Problem`. It has `t_d` present in `u(t-t_d)`. The `ConstraintCollocator` then looks through this eom, finds all expressions of the form U(time_symbol +/- symbol) and substitutes them with U_shift(time_symbol.

moorepants commented on opty/tests/test_direct_collocation.py:
```diff
@@ -1566,3 +1566,51 @@ def test_for_algebraic_eoms():
+        self.state_symbols = (x_1, x_2)
+        self.constant_symbols = (a_0, a_1)
+        self.constant_values = (1, np.sqrt(2))
+        self.timeshift_inputs = {u_shift(t): (u(t - t_d), t_d)}
```
Why did you choose adding a new attribute here, instead of passing u_shift - u(t - t_d) directly to instance constraints?

chris-konrad replied:
The `ConstraintCollocator` automatically appends instance constraints representing u_shift(t) - u(t - t_d) to the other instance constraints, I just didn't write a test for this yet. The attribute `ConstraintCollocator.timeshift_trajectory_substitutes` that I check in this test is just a map from the substitutes to the original trajectory and the timeshift parameter that I occasionally need throughout the code to look up which functions and parameters relate to the timeshift.

chris-konrad commented:
> > This assumes the following layout of the Jacobian. Can you confirm that this is correct, @moorepants?
> 
> Are you treating τ the same as any unknown parameter? i.e. is it in the list of `unknown_parameters`? This would have some bearing on the construction of the jacobian (order of rows).

Thats what I do. Indeed there seems to be a problem in the Jacobian index function related to this. I will have a look.

chris-konrad commented:
Added unittest to check eom discretization and the jacobian matrix. Should be fine, but IPOPT/cyipopt still complains about invalid jacobian row indices: 

```
This is Ipopt version 3.14.16, running with linear solver MUMPS 5.7.1.

Exception of type: INVALID_TNLP in file "D:\bld\ipopt_1717602782538\work\src\Interfaces\IpTNLPAdapter.cpp" at line 1100:
 Exception message: eval_jac_g returned false in GetSpaces

EXIT: Some uncaught Ipopt exception encountered.
ERROR:cyipopt:b'Invalid row indices returned from jacobianstructure'
Some uncaught Ipopt exception encountered.
objective value:    0.000000
identified T_0: 1.000000
true T_0:       10.000000

```

moorepants commented on opty/direct_collocation.py:
```diff
@@ -1680,6 +1958,9 @@ def constraints(free):
                                 time_interval)
 
+            if all_specified.shape[0] == 1:
+                all_specified.squeeze()
```
Should be safe to call squeeze() without the if statement, as it will only affect arrays with an empty dimension.

Peter230655 commented:
I may have found a simple hack to do this:
In the time delayed driving force $u(t - \tau)$ I replaced t with a variable T, and in the eom I set $\dfrac{d}{dt} T(t) = 1.0$

moorepants commented:
Yes, this sounds like a useful example to have.

moorepants commented:
Are you doing `u(T(t)  - tau)`?

Peter230655 commented:
> Are you doing `u(T(t) - tau)`?

Yes, this is exactly what I do. So far looks o.k.

moorepants commented:
I don't think the internals of opty support anything but `t` in the function `Function('u')(t)`.

Peter230655 commented:
Of course only you know this best!
I used ``sm.sin(T(t) - $\tau)$. My simple thinking was this: opty does accept functions of state variables and treats them correctly.
So if I introduce a state valiable with the smae numerical values as ``t`` it might work.
So far looks o.k., but I will play around more.
Actually I used torque = $F \cdot \sin(\omega \cdot (T(t) - \tau)$ and let it estimate F, $\omega, \tau$.

moorepants commented:
` s i n ( T ( t ) − τ ) ` will work, but not if the function is a state variable, e.g. `x(T(t) - t)` or input variable `r(T(t) - t)`. I think all of the instance constraint code only works with `x(t)` or `r(t)` and should break (or give incorrect results) otherwise.

moorepants commented:
Feel free to prove me wrong though!

Peter230655 commented:
I do not think, you are wrong! :-)
I also do not think, a state variable will work with anything but ``t``, but this is not what I need - if I understood Chris' problem correctly.

moorepants commented:
>  I also do not think, a state variable will work with anything but t, but this is not what I need - if I understood Chris' problem correctly.

Christoph precisely needs an unknown input to have a delay, so `r(t - tau)` instead of `r(t)`.

Peter230655 commented:
Then I missunderstood, was too easy to be true! 😊😊
So is opty expected to estimate the delay and what function r is?

Peter230655 commented:
> > I also do not think, a state variable will work with anything but t, but this is not what I need - if I understood Chris' problem correctly.
> 
> Christoph precisely needs an unknown input to have a delay, so `r(t - tau)` instead of `r(t)`.

@moorepants This has been dormant for a while. I am still confused: @chris-konrad wants to estimate the unknown delay of an unknown input function, so estimate $u(t - \tau)$ where both u and $\tau$ are unknown. How is this possible in principle?
$v(t) := u(t - \tau)$ would also work. so how could opty distinguish between the functions $u$ and $v$ ?
Not urgent, I just stumbled across this looking for something else.

chris-konrad commented:
I want to find the timeshift tau of an otherwise known function u(t-tau). I know the shape of the input, just not the timeshift. 

To make it a little more concrete: I have a scenario where a dynamic system is exposed to an input step. I know the input level it started with and I know the input level it ended with, but I don't know when the jump occured. 

It is the classic case of trying to find an unknown input delay. This could be for example a network latency, some fixed processing delay or a human reaction time.
